src/thread/thread_looper.cpp

```diff
@@ -115,4 +135,69 @@ namespace cubthread
     return m_was_woken_up;
   }
 
+  looper::delta_time
+  looper::get_wait_for (void)
+  {
+    delta_time execution_time;
+
+    if (m_start_execution_time == std::chrono::system_clock::time_point ())
+      {
+	// assume execution_time is 0 (zero) for first run
+	// since m_start_execution_time is not yet registered
+	execution_time = delta_time (0);
+      }
+    else
+      {
+	execution_time = std::chrono::system_clock::now () - m_start_execution_time;
+      }
+
+    delta_time wait_for;
+    delta_time period = m_periods[m_period_index];
+
+    // compute task execution time
+    if (execution_time < period)
+      {
+	wait_for = period - execution_time;
+      }
+    else
+      {
+	// if execution_time is greater than the period then immediately execute task without sleeping
+	wait_for = delta_time (0);
+      }
+
+    assert (wait_for <= period);
+
+    return wait_for;
+  }
+
+  void
+  looper::setup_period (void)
+  {
+    if (!m_period_function || m_wait_pattern == wait_pattern::INCREASING_PERIODS)
```

Comment:
This condition looks a little bit weird. Maybe it is better to rewrite put_to_sleep/setup_period to clarify the logic.

First thing in put_to_sleep is to decide what to do next. We used to have three behaviors, now we have four:
  1. infinite waits on each iteration.
  2. fixed period on each iteration
  3. increasing periods on each iteration.
  4. [NEW] based on a function provided at construction.

After checking behavior & status, there are three possible decisions: infinite wait, no wait, or timed wait. No wait can be considered timed wait of zero, thus the decisions are reduced to two.

I would like to propose to use different setup_period functions in all cases, that output two variables: a boolean to decide between infinite waits and timed waits and a time for timed wait.

Instead of "wait patterns" enum, we can have private wait pattern functions:

```c
void
setup_infinite_wait (bool & is_timed_wait, duration & delta_time)
{
  is_timed_wait = false;
}

void
setup_fixed_waits (...

void
setup_increasing_waits (...

void
put_sleep (waiter &)
{
  ...

  m_setup_period (is_timed_wait, delta_time);
  if (is_timed_wait)
    // subtract execution time and wait
  else
    // infinite wait
}
```

What do you think?